crates/iroha_executor_data_model/src/permission.rs

```diff
@@ -133,121 +121,109 @@ pub mod asset {
     }
 
     permission! {
-        pub struct CanUnregisterUserAsset {
-            pub asset: AssetId,
+        pub struct CanMintAssetsWithDefinition {
+            pub asset_definition: AssetDefinitionId,
         }
     }
 
     permission! {
-        pub struct CanBurnAssetWithDefinition {
+        pub struct CanBurnAssetsWithDefinition {
             pub asset_definition: AssetDefinitionId,
         }
     }
 
     permission! {
-        pub struct CanBurnUserAsset {
-            pub asset: AssetId,
+        pub struct CanTransferAssetsWithDefinition {
+            pub asset_definition: AssetDefinitionId,
         }
     }
 
     permission! {
-        pub struct CanMintAssetWithDefinition {
-            pub asset_definition: AssetDefinitionId,
+        pub struct CanRegisterAsset {
+            pub owner: AccountId,
         }
     }
 
     permission! {
-        pub struct CanMintUserAsset {
+        pub struct CanUnregisterAsset {
             pub asset: AssetId,
         }
     }
 
     permission! {
-        pub struct CanTransferAssetWithDefinition {
-            pub asset_definition: AssetDefinitionId,
+        pub struct CanMintAsset {
+            pub asset: AssetId,
         }
     }
 
     permission! {
-        pub struct CanTransferUserAsset {
+        pub struct CanBurnAsset {
             pub asset: AssetId,
         }
     }
 
     permission! {
-        pub struct CanSetKeyValueInUserAsset {
+        pub struct CanTransferAsset {
             pub asset: AssetId,
         }
     }
 
     permission! {
-        pub struct CanRemoveKeyValueInUserAsset {
+        pub struct CanModifyAssetMetadata {
```

Comment:
On the one hand, would be great to keep these primitive and build on top of them (maybe through some concept of permission bundles). On the other, I can't come up with an example where you would need to be able to set and remove key-values separately. Maybe a producer-consumer type of situation. It's difficult to tell without user feedback, I think for the MVP this should be fine.

(Also, it's interesting that we lean heavily into generic structs with ISI (and don't really reap the benefits), but permissions/queries are all flat structs.)

Reply:
> On the one hand, would be great to keep these primitive and build on top of them (maybe through some concept of permission bundles). On the other, I can't come up with an example where you would need to be able to set and remove key-values separately. Maybe a producer-consumer type of situation. It's difficult to tell without user feedback, I think for the MVP this should be fine.

I think having 2 separate permissions doesn't make sense since both operations are write operations. If you give a permission to `SetKeyValue` that's almost the same as if you gave the permission to `RemoveKeyValue` since it destroys the previous value so you're not really getting more protection

> (Also, it's interesting that we lean heavily into generic structs with ISI (and don't really reap the benefits), but permissions/queries are all flat structs.)

I was thinking the same. Would be worth exploring